### PR TITLE
Feat/node/start from refresh token

### PR DESCRIPTION
**Pull request by NSeydoux**
This changes the internal API of the login function: in the case of a refresh flow when the refresh token is provided (and later on in the client credential flow), getting the access token happens directly in the back channel, and no front channel interaction is required, which means no redirection happens. So far, our `login` function only ever returned `void`, because it
was assumed that we were going through the auth code flow or the implicit flow, and that the refresh flow only happened after these once the fetch function is already instantiated.

Now, the login function can return the result of the token exchange if appropriate, and undefined if we expect the login to be completed after a redirection. All the login handlers are aligned on the new signatures . If necessary, their implementation is slightly changed to comply with the new expected return type (e.g. functions which were void now return undefined).

- [X] All acceptance criteria are met.
- [x] The changelog has been updated, if applicable.
- [X] New functions/types have been exported in `index.ts`, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: packages/node/src/login/AggregateLoginHandler.ts [hunk 1]
```diff
@@ -32,14 +32,18 @@ import {
...
 
 /**
  * @hidden
  */
 @injectable()
 export default class AggregateLoginHandler
-  extends AggregateHandler<[ILoginOptions], void>
+  extends AggregateHandler<
+    [ILoginOptions],
+    (ISessionInfo & { fetch: typeof fetch }) | undefined
```
Comment (pmcb55) [inline packages/node/src/login/AggregateLoginHandler.ts:45]: Isn't this just the type `LoginResult`...?

File: packages/node/src/login/oidc/AggregateOidcHandler.ts [hunk 2]
```diff
@@ -32,14 +32,19 @@ import {
...
 
 /**
  * @hidden
  */
 @injectable()
 export default class AggregateOidcHandler
-  extends AggregateHandler<[IOidcOptions], void>
+  extends AggregateHandler<
+    [IOidcOptions],
+    (ISessionInfo & { fetch: typeof fetch }) | undefined
```
Comment (pmcb55) [inline packages/node/src/login/oidc/AggregateOidcHandler.ts:46]: Ditto - i.e. isn't this just the type `LoginResult`...?

File: packages/node/src/login/oidc/OidcLoginHandler.ts [hunk 3]
```diff
@@ -67,24 +69,26 @@ export default class OidcLoginHandler implements ILoginHandler {
...
 
   async canHandle(options: ILoginOptions): Promise<boolean> {
-    return hasIssuer(options) && hasRedirectUrl(options);
+    return hasIssuer(options);
   }
 
-  async handle(options: ILoginOptions): Promise<void> {
+  async handle(
+    options: ILoginOptions
+  ): Promise<(ISessionInfo & { fetch: typeof fetch }) | undefined> {
```
Comment (pmcb55) [inline packages/node/src/login/oidc/OidcLoginHandler.ts:77]: Ditto - i.e. isn't this just the type `LoginResult`...?

File: packages/node/src/login/oidc/OidcLoginHandler.ts [hunk 4]
```diff
@@ -67,24 +69,26 @@ export default class OidcLoginHandler implements ILoginHandler {
...
       );
     }
-    if (!hasRedirectUrl(options)) {
-      throw new ConfigurationError(
-        `OidcLoginHandler requires a redirect URL: missing property 'redirectUrl' in ${JSON.stringify(
-          options
-        )}`
-      );
-    }
+    // if (!hasRedirectUrl(options)) {
```
Comment (pmcb55) [inline packages/node/src/login/oidc/OidcLoginHandler.ts:85]: Is this block of commented code left in intentionally, or should it just be removed...?

File: packages/node/src/login/oidc/oidcHandlers/AuthorizationCodeWithPkceOidcHandler.ts [hunk 5]
```diff
@@ -75,6 +76,8 @@ export default class AuthorizationCodeWithPkceOidcHandler
       response_type: "code",
       redirect_uri: oidcLoginOptions.redirectUrl,
       code_challenge_method: "S256",
+      // The offline_access scope asks the provider to issue a refresh token
```
Comment (pmcb55) [inline packages/node/src/login/oidc/oidcHandlers/AuthorizationCodeWithPkceOidcHandler.ts:79]:
```suggestion
      // The offline_access scope asks the provider to issue a refresh token.
```

File: packages/node/src/login/oidc/oidcHandlers/RefreshTokenOidcHandler.ts [hunk 6]
```diff
@@ -30,18 +30,109 @@
...
-    throw new NotImplementedError("RefreshTokenOidcHandler handle");
+  async handle(oidcLoginOptions: IOidcOptions): Promise<LoginResult> {
+    if (!(await this.canHandle(oidcLoginOptions))) {
+      throw new Error(
+        `RefreshTokenOidcHandler cannot handle the provided options, missing one of 'refreshToken', 'clientId' in: ${JSON.stringify(
+          oidcLoginOptions
+        )}`
+      );
+    }
+    // throw new NotImplementedError("RefreshTokenOidcHandler handle");
```
Comment (pmcb55) [inline packages/node/src/login/oidc/oidcHandlers/RefreshTokenOidcHandler.ts:81]: Remove this...?

**Review by pmcb55 — APPROVED** [review]
LGTM (just very minor comments).

File: packages/node/src/login/AggregateLoginHandler.ts (hunk 1 above)
Comment (NSeydoux) [reply packages/node/src/login/AggregateLoginHandler.ts]: Correct, I named the type as a shorthand when I realized I was using it in a lot of places, but I didn't look for it everywhere it seems. Fixed in 6ee8aac0c890e90d7e3af0fdf265de1327bb0a53

File: packages/node/src/login/oidc/AggregateOidcHandler.ts (hunk 2 above)
Comment (NSeydoux) [reply packages/node/src/login/oidc/AggregateOidcHandler.ts]: Fixed in 6ee8aac0c890e90d7e3af0fdf265de1327bb0a53

File: packages/node/src/login/oidc/OidcLoginHandler.ts (hunk 3 above)
Comment (NSeydoux) [reply packages/node/src/login/oidc/OidcLoginHandler.ts]: Fixed in 6ee8aac0c890e90d7e3af0fdf265de1327bb0a53

File: packages/node/src/login/oidc/OidcLoginHandler.ts (hunk 4 above)
Comment (NSeydoux) [reply packages/node/src/login/oidc/OidcLoginHandler.ts:86]: I'll clean this up as part of the housekeeping: this code should be moved elsewhere. I added a TODO note: 18b481a745bd001ffc8df6c8d436262a59968fb9

File: packages/node/src/login/oidc/oidcHandlers/RefreshTokenOidcHandler.ts (hunk 6 above)
Comment (NSeydoux) [reply packages/node/src/login/oidc/oidcHandlers/RefreshTokenOidcHandler.ts]: 6591a4233f6ba313670d3f63c0de2283bd67dbde

**Pull request merged**